datanode/snapshot/service_load_snapshot.go

```diff
@@ -0,0 +1,433 @@
+package snapshot
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"code.vegaprotocol.io/vega/datanode/entities"
+	"code.vegaprotocol.io/vega/datanode/snapshot/orders"
+	"code.vegaprotocol.io/vega/datanode/sqlstore"
+	"github.com/georgysavva/scany/pgxscan"
+	"github.com/jackc/pgconn"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+)
+
+var ErrNoCurrentStateSnapshotFound = errors.New("no current state snapshot found")
+
+type Conn interface {
+	Exec(ctx context.Context, sql string, arguments ...interface{}) (pgconn.CommandTag, error)
+	Query(ctx context.Context, sql string, args ...interface{}) (pgx.Rows, error)
+	QueryRow(ctx context.Context, sql string, args ...interface{}) pgx.Row
+}
+
+func (b *Service) LoadAllAvailableHistory(ctx context.Context) (int64, int64, error) {
+	oldestHistoryBlock, lastBlock, err := GetOldestHistoryBlockAndLastBlock(ctx, b.connConfig, b.blockStore)
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed to get oldest history block and last block:%w", err)
+	}
+
+	chainID, currentStateSnapshot, contiguousHistory, err := GetAllAvailableHistory(b.snapshotsPath, oldestHistoryBlock, lastBlock)
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed to get all available history: %w", err)
+	}
+
+	if datanodeHasData(oldestHistoryBlock) {
+		if len(chainID) > 0 {
+			datanodeChainID, err := b.chainService.GetChainID()
+			if err != nil {
+				return 0, 0, fmt.Errorf("failed to get datanode chain id:%w", err)
+			}
+
+			if len(datanodeChainID) > 0 {
+				if chainID != datanodeChainID {
+					return 0, 0, fmt.Errorf("available history chain id %s does not match datanodes existing chain id %s", chainID, datanodeChainID)
+				}
+			}
+		}
+
+		b.log.Infof("snapshotting all datanode data")
+		meta, err := b.CreateSnapshot(ctx, chainID, oldestHistoryBlock.Height, lastBlock.Height)
+		if err != nil {
+			return 0, 0, fmt.Errorf("failed to snapshot all datanode data: %w", err)
+		}
+
+		b.log.Infof("snapshotted all datanode data from height %d to %d", meta.HeightFrom, meta.HeightTo)
+	}
+
+	allSnapshotData := appendCurrentStateSnapshotOntoHistory(contiguousHistory, currentStateSnapshot)
+
+	start := time.Now()
+	b.log.Infof("restoring data node from snapshot data %+q", allSnapshotData)
+
+	b.log.Infof("creating database")
+	err = sqlstore.RecreateVegaDatabase(ctx, b.log, b.connConfig)
+	if err != nil {
+		return 0, 0, fmt.Errorf("failed to create vega database: %w", err)
+	}
```

Comment:
I do favour this form usually when the err is used only in one context to scope it
```suggestion
	if err = sqlstore.RecreateVegaDatabase(ctx, b.log, b.connConfig); if err != nil {
		return 0, 0, fmt.Errorf("failed to create vega database: %w", err)
	}
```